File: CPP/linear_search.cpp
#include <bits/stdc++.h>
using namespace std;

int linear_search(int key, int arr[], int arr_size)
{
    for (int i = 0; i < arr_size; i++)
    {
        if (arr[i] == key)
            return (i + 1);
    }

    return -1;
}

int main()
{

    //fast input-output
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    // Input Size and Array Elements
    int arr_size;
    cin >> arr_size;
    int arr[arr_size];
    for (int i = 0; i < arr_size; i++)
        cin >> arr[i];

    // Input the element to be searched
    int key;
    cin >> key;

    int pos = linear_search(key, arr, arr_size);
    if (pos > 0)
        cout << "ELEMENT FOUND AT POSITION " << pos;
    else
        cout << "ELEMENT NOT FOUND !";

    return 0;
}
/*
Linear search implementation in C++
Input: Size of array , Array of integers and the value to be searched in the array
Output: Returns position of the element if it's present else '-1'
----------------------------------------
Sample Input 1:
6
[11, 10, 6, 7, 8, 19]
6
Output:
ELEMENT FOUND AT POSITION 3


Sample Input 2:
6
[10,33,23,22,11,1]
5
Output:
ELEMENT NOT FOUND !
----------------------------------------
Time Complexity:
Worst Case = O(n)
Best Case(when the first element is the desired element) = O(1)
Space Complexity: O(1)
*/

File: Go/linearSearch.go
//Linear Search
/*
Example 1:
Input:9
Output:2

Example 2:
Input:1100
Output:-1

Example 3:
Input:1100
Output:7

*/

/*
Time Complexity of Linear Search is O(n) where n is length of array.
Space Complexity of Linear Search is O(1).

*/
package main

import "fmt"

// function linearSearch return posstion of the element if present otherwise returns -1.
func linearSearch(array []int, num int) int {
	for i := 0; i < len(array); i++ {
		if array[i] == num {

			return i
		}
	}
	return -1
}

func main() {
	items := []int{45, 100, 9, 1, 2, 20, 31, 63, 70}
	ans := linearSearch(items, 9)
	//ans := linearSearch(items, 1110)
	if ans == -1 {
		fmt.Println("Number is not present in array")
	} else {
		fmt.Printf("Number is present at the location :%d", ans)
	}
}
